File: src/Butil/wwwroot/butil.js
var butil = (function () {
    let _dotnetObj = null;

    return {
        init,
        register
    };

    function init(dotnetObj) {
        _dotnetObj = dotnetObj;
    }

    function register(elementName, eventName, dotnetMethodName, dotnetMethodId, selectedMembers, options) {
        const handler = e => {
            _dotnetObj.invokeMethodAsync(dotnetMethodName, dotnetMethodId, selectedMembers && select(e, selectedMembers));
        };

        window[elementName].addEventListener(eventName, handler, options);
    }

    function select(source, members) {
        return members.reduce((pre, cur) => pre[cur] = source[cur], {});
    }
}());

File: src/Butil/Butil.cs
using System;
using Microsoft.JSInterop;

namespace Butil;

public static class Butil
{
    private static bool _isInitialized;
    private static IJSRuntime _js = default!;

    public static void Init(IJSRuntime jsRuntime)
    {
        if (_isInitialized) return;

        _isInitialized = true;
        _js = jsRuntime;
    }

    internal static void AddEventListener(string elementName, string eventName, string dotnetMethodName, Guid dotnetListenerId, string[] selectedMembers, object options = null)
    {
        _js.InvokeVoidAsync("butil.addEventListener", elementName, eventName, dotnetMethodName, dotnetListenerId, selectedMembers, options);
    }

    internal static void RemoveEventListener(string elementName, string eventName, Guid dotnetListenerId, object options = null)
    {
        _js.InvokeVoidAsync("butil.removeEventListener", elementName, eventName, dotnetListenerId, options);
    }
}
